compose/ui/ui-text/src/skikoMain/kotlin/androidx/compose/ui/text/platform/ParagraphLayouter.skiko.kt

```diff
@@ -61,49 +68,99 @@ class ParagraphLayouter(
-        width: Float = this.width,
-        maxLines: Int = builder.maxLines,
-        ellipsis: String = builder.ellipsis,
-        color: Color = builder.textStyle.color,
-        shadow: Shadow? = builder.textStyle.shadow,
-        textDecoration: TextDecoration? = builder.textStyle.textDecoration,
-    ): Paragraph {
+    fun setParagraphStyle(
+        maxLines: Int,
+        ellipsis: String
+    ) {
+        if (builder.maxLines != maxLines ||
+            builder.ellipsis != ellipsis
+        ) {
+            builder.maxLines = maxLines
+            builder.ellipsis = ellipsis
+            paragraphCache = null
+        }
+    }
+
+    fun setTextStyle(
+        color: Color,
+        shadow: Shadow?,
+        textDecoration: TextDecoration?
+    ) {
         val actualColor = color.takeOrElse { builder.textStyle.color }
-        if (
-            builder.maxLines != maxLines ||
-            builder.ellipsis != ellipsis ||
-            builder.textStyle.color != actualColor ||
+        if (builder.textStyle.color != actualColor ||
             builder.textStyle.shadow != shadow ||
             builder.textStyle.textDecoration != textDecoration
         ) {
-            this.width = width
-            builder.maxLines = maxLines
-            builder.ellipsis = ellipsis
             builder.textStyle = builder.textStyle.copy(
                 color = actualColor,
                 shadow = shadow,
                 textDecoration = textDecoration
             )
-            para = builder.build()
-            para.layout(width)
-        } else if (this.width != width) {
-            this.width = width
-            para.layout(width)
+            paragraphCache = null
         }
-        return para
     }
+
+    @ExperimentalTextApi
+    fun setTextStyle(
+        brush: Brush?,
+        brushSize: Size,
+        alpha: Float,
+        shadow: Shadow?,
+        textDecoration: TextDecoration?
+    ) {
+        val actualSize = builder.brushSize
+        if (builder.textStyle.brush != brush ||
+            actualSize == null ||
+            !actualSize.width.sameValueAs(brushSize.width) ||
+            !actualSize.height.sameValueAs(brushSize.height) ||
+            !builder.textStyle.alpha.sameValueAs(alpha) ||
+            builder.textStyle.shadow != shadow ||
+            builder.textStyle.textDecoration != textDecoration
+        ) {
+            builder.textStyle = builder.textStyle.copy(
+                brush = brush,
+                alpha = alpha,
+                shadow = shadow,
+                textDecoration = textDecoration
+            )
+            builder.brushSize = brushSize
+            paragraphCache = null
+        }
+    }
+
+    fun setDrawStyle(drawStyle: DrawStyle?) {
+        // TODO Implement applying DrawStyle
+    }
+
+    fun layoutParagraph(width: Float): Paragraph {
+        val paragraph = paragraphCache
+        return if (paragraph != null) {
+            if (!this.width.sameValueAs(width)) {
```

Comment:
We can't reuse the same paragraph when we call `layoutParagraph` with a different width. ParagraphLayouter is created by SkiaParagraphIntrinsics, that can be shared between 2 SkiaParagraph.

We had [this cache](https://github.com/JetBrains/compose-multiplatform-core/commit/b1161bd7e486da15691240589f911d3467557198) and had a bug because of it.

Reply:
Actually, that is how it was before the PR. Not sure, but we can have a bug, when users try to get metrics (will test it later)

Reply:
And actually, we don't use the same instance after that PR. I forgot that I fixed it by this code :)
```
    private var layouter: ParagraphLayouter? = newLayouter()

    fun layouter(): ParagraphLayouter {
        val layouter = this.layouter ?: newLayouter()
        this.layouter = null
        return layouter
    }
```

The name deceived me, better to name it as `createLayouter`, even if internally we just reuse it the first time. We cache `layouter` only to know `minIntrinsicWidth`.